src/GraphQL/API/Enum.hs

```diff
@@ -0,0 +1,101 @@
+  genericEnumValues :: [Either NameError Name]
+  genericEnumFromValue :: GValue.Value -> Either Text (f p)
+  genericEnumToValue :: f p -> GValue.Value
+
+instance forall conName m p f nt.
+  ( KnownSymbol conName
+  , KnownSymbol m
+  , KnownSymbol p
+  , GenricEnumValues f
+  ) => GenricEnumValues (M1 D ('MetaData conName m p nt) f) where
+  genericEnumValues = genericEnumValues @f
+  genericEnumFromValue v@(GValue.ValueEnum _) = M1 <$> genericEnumFromValue v
+  genericEnumFromValue x = Left ("Not an enum: " <> show x)
+  genericEnumToValue (M1 gv) = genericEnumToValue gv
+
+instance forall conName f p b.
+  ( KnownSymbol conName
+  , GenricEnumValues f
+  ) => GenricEnumValues (C1 ('MetaCons conName p b) U1 :+: f) where
+  genericEnumValues = let name = nameFromSymbol @conName in name:(genericEnumValues @f)
+  genericEnumFromValue v@(GValue.ValueEnum vname) =
+    case nameFromSymbol @conName of
+      Right name -> if name == vname
+                    then L1 <$> Right (M1 U1)
+                    else R1 <$> genericEnumFromValue v
+      Left x -> Left ("Not a valid enum name: " <> show x)
+  genericEnumFromValue _ = panic "This case should have been caught at top-level. Please file a bug."
+  genericEnumToValue (L1 _) =
+    let Right name = nameFromSymbol @conName
+    in GValue.ValueEnum name
+  genericEnumToValue (R1 gv) = genericEnumToValue gv
+
+instance forall conName p b. (KnownSymbol conName) => GenricEnumValues (C1 ('MetaCons conName p b) U1) where
+  genericEnumValues = let name = nameFromSymbol @conName in [name]
+  genericEnumFromValue (GValue.ValueEnum vname) =
+    case nameFromSymbol @conName of
+      Right name -> if name == vname
+                    then Right (M1 U1)
+                    else Left ("Not a valid enum name: " <> show vname)
+      Left x -> Left ("Not a valid enum name: " <> show x)
+  genericEnumFromValue _ = panic "This case should have been caught at top-level. Please file a bug."
+  genericEnumToValue (M1 _) =
+    let Right name = nameFromSymbol @conName
+    in GValue.ValueEnum name
+
+-- TODO(tom): better type errors using `n`. Also type errors for other
+-- invalid constructors.
+instance forall conName p b sa sb.
+  ( TypeError ('Text "Constructor not unary: " ':<>: 'Text conName)
+  , KnownSymbol conName
+  ) => GenricEnumValues (C1 ('MetaCons conName p b) (S1 sa sb)) where
+  genericEnumValues = notImplemented
+  genericEnumFromValue = notImplemented
+  genericEnumToValue = notImplemented
+
+instance forall conName p b sa sb f.
+  ( TypeError ('Text "Constructor not unary: " ':<>: 'Text conName)
+  , KnownSymbol conName
+  ) => GenricEnumValues (C1 ('MetaCons conName p b) (S1 sa sb) :+: f) where
+  genericEnumValues = notImplemented
+  genericEnumFromValue = notImplemented
+  genericEnumToValue = notImplemented
+
+
+-- | For each enum type we need 1) a list of all possible values 2) a
+-- way to serialise and 3) deserialise.
+class GraphQLEnum a where
+  enumValues :: [Either NameError Name]
```

Comment:
If we make users use  `makeName` we're leaning on our users to do all the error checking just for constructing the list of `Name` (which can be done statically most of the time). They may then be tempted to use `unsafeMakeName` to avoid the extra code, and then we have to do error checking in `API` because we can't rely on the `Name` contract.

Not sure what the right solution here is but making all users write sth like 

```haskell
enumValues = let Right x = sequenceA [makeName "A", makeName "B"] in x
```

doesn't seem great either..

not changed for now, any other ideas?